### ISSUE TITLE
update-reset does not account for shallow clones

### ISSUE THREAD

**Issue by nebhale** [Homebrew/brew#7795]
- [X] ran `brew update` and can still reproduce the problem?
- [X] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [X] ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

After your quick turn around on #7746 (thanks for that!), I immediately dove into changing a couple of taps to use `main` instead of `master`.  When running `brew upgrade` I ran into some issues, so I did the more surgical method referred to [here](https://github.com/Homebrew/brew/issues/7746#issuecomment-647727185).

This surgical method had the same same issue which was that `git remote set-head origin --auto` failed (`error: Not a valid ref: refs/remotes/origin/main`) because the tap was a shallow clone and the ref returned from `--auto` did not exist in that shallow clone.

## What happened (include command output)

<details>
  <summary>Command output</summary>
  <pre>
➜  ~ brew update-reset $(brew --repo buildpack/tap)         
==> Fetching /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...
error: Not a valid ref: refs/remotes/origin/main

==> Resetting /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...
Branch 'master' set up to track remote branch 'master' from 'origin'.
Reset branch 'master'
Your branch is up to date with 'origin/master'.
  </pre>
</details>

<details>
  <summary>Command output (debug)</summary>
  <pre>
➜  ~ HOMEBREW_DEBUG=true brew update-reset $(brew --repo buildpack/tap)
+ [[ -z /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap ]]
+ for DIR in '"${REPOS[@]}"'
+ [[ -d /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap/.git ]]
+ cd /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap
+ ohai 'Fetching /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...'
+ [[ -n '' ]]
+ [[ -t 1 ]]
+ [[ -z '' ]]
+ echo -e '\033[34m==>\033[0m \033[1mFetching /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...\033[0m'
==> Fetching /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...
+ git fetch --force --tags origin
+ /usr/local/Homebrew/Library/Homebrew/shims/scm/git fetch --force --tags origin
+ git remote set-head origin --auto
+ /usr/local/Homebrew/Library/Homebrew/shims/scm/git remote set-head origin --auto
error: Not a valid ref: refs/remotes/origin/main
+ echo
+ ohai 'Resetting /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...'
+ [[ -n '' ]]
+ [[ -t 1 ]]
+ [[ -z '' ]]
+ echo -e '\033[34m==>\033[0m \033[1mResetting /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...\033[0m'
==> Resetting /usr/local/Homebrew/Library/Taps/buildpack/homebrew-tap...
++ git symbolic-ref refs/remotes/origin/HEAD
++ /usr/local/Homebrew/Library/Homebrew/shims/scm/git symbolic-ref refs/remotes/origin/HEAD
+ head=refs/remotes/origin/master
+ head=master
+ git checkout --force -B master origin/HEAD
+ /usr/local/Homebrew/Library/Homebrew/shims/scm/git checkout --force -B master origin/HEAD
Branch 'master' set up to track remote branch 'master' from 'origin'.
Reset branch 'master'
Your branch is up to date with 'origin/master'.
+ echo
+ exit 0
  </pre>
</details>

## What you expected to happen

I expected the tap to be updated to the new default branch, as #7746 was aiming for.

## Step-by-step reproduction instructions (by running `brew` commands)

1. `brew tap <tap-name>`
1. Update the default branch of the tap repository, deleting the previous default branch
1. `brew update-reset $(brew --repo <tap-name>)`

## Output of `brew config` and `brew doctor` commands

<pre>
➜  ~ brew config                               
HOMEBREW_VERSION: 2.4.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: a89b56598bf2334bf53fb9e55694551ecf680916
Last commit: 13 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 9f77ca3d4e7b99f35479844a2e6157822e95a489
Core tap last commit: 51 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_EDITOR: subl -w
HOMEBREW_MAKE_JOBS: 8
CPU: octa-core 64-bit skylake
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
Clang: 11.0 build 1103
Git: 2.27.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 14.0.1+8, 11.0.7+10-LTS, 11.0.7, 1.8.0_252-b09
macOS: 10.15.5-x86_64
CLT: 1103.0.32.62
Xcode: 11.5
➜  ~ brew doctor
Your system is ready to brew.

</pre>

**Comment by sjackman** [comment]
A PR to fix this issue is welcome if you'd like to open one. If you have a halfway working fix, and want to discuss it, you can open a draft PR, and we can chat there.

**Comment by nebhale** [comment]
I don't have a PR for it, nor really the expertise in Homebrew's internals to create one.  This was more me raising a flag that the two changes that went in earlier for this root issue don't appear to work in (what I believe is) a common scenario.

Perhaps it's because I use a `Brewfile` to setup my development machines, but I've never knowingly tapped with the `--shallow` flag and yet all of my taps are shallow clones today.  If you tell me that this is an aberration, I'll attempt to update my `Brewfile` to do full clones to get back to "typical" behavior and call it day.

**Comment by sjackman** [comment]
The shallow clone is I believe a red herring. The issue is that renaming a branch from `master` to `main` does not currently work.

**Comment by nebhale** [comment]
Ah, interesting!  I didn’t have any full clones to test it with so assumed it was a problem on my end.  I’ll do some additional testing with non-Homebrew Git repos and see if I can figure a different set of commands that can do what you’re aiming for.  It’s a damn hard problem and I was really surprised to see how quickly you all had knocked it out.

Let me say thanks again for attempting to add the feature so quickly; it really makes a difference to the broader community making this change at the moment.

**Comment by MikeMcQuaid** [comment]
Closing as duplicate of the closed and pinned https://github.com/Homebrew/brew/issues/7746

**Comment by nebhale** [comment]
@sjackman I've just confirmed this morning that `update-reset` works as designed for non-shallow clones.  So the failure is related to the restrictions on shallow clones.

@MikeMcQuaid I think this issue should be re-opened as a non-duplicate as it is a materially different issue than #7746.  In essence, #7746 has been solved for the deep clone, but there is still a bug when dealing with shallow clones.
